text/0022-gas-spot-market.md

```diff
@@ -0,0 +1,215 @@
+rule.
+
+### What does the user see?
+
+The UI change for the user is the introduction (possibly by renaming) of a "willingness to pay" parameter and a "spot gas reservation" parameter. The UX
+becomes somewhat more complex, since it is up to the user to determine the gas requirements. Because of our consensus-before-execution model, we are unable
+to allocate gas on the fly. This means that to avoid underutilization of block gas, we must charge for the entire reserved amount. It is expected that the users, and particularly high WTP users, will make their best effort to estimate gas
+requirements offline, or on our testnet, before submitting the deploys to the live network. We expect that, eventually, the community will develop tooling to make this estimation easier.
+Some of the potential solutions may be large dApps contributing developer time and nodes to maintain a robust testnet that shadows complex contracts deployed on the live chain, adaptation of
+our testing tools and application of empirical methods to observed gas use patterns.
+
+### What does the protoblock proposer see?
+
+The protoblock proposer, here a personification of the block proposer component, sees a deploy buffer, as usual. However, the block proposer will now be hardcoded to follow
+the simple greedy rule we outlined above, alongside with balance checks to prevent attacks by "forgetting one's wallet at home" (we will refer to this, more formally, as the "nothing-in-purse problem").
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### Chainspec parameters
+
+There are several natural parameters that govern the process we outlined. While it would reduce the scope of proposed work to leave these hardcoded, the optimal values cannot be determined in advance and will change as the platform grows, making it imperative that these parameters be included in the chainspec.
+
+#### General gas limit
+
+The limit on spot gas reservations by general deploys.
+
+#### WASM-less transfer dedicated gas limit
+
+The limit on spot gas reservations of WASM-less deploys, naturally stated as a multiple of desired number of such deploys multiplied by their (fixed) gas reservation. 
+
+### Deploy channels
+
+We have already mentioned the concept of a dedicated gas limit for WASM-less transfers in our description of the proposed chainspec parameters. To avoid crowding out WASM-less transfers, which are expected to be critical for platform health and public perception, there should be a separate "lane" for such transfers and this parameter specifies the width of this "lane." Selection among WASM-less transfers would work the same as it does for general deploys, but with sorting by WTP alone, since the gas reservation is fixed for these. Moreover, WASM-less transfers that are not selected for inclusion would be eligible for inclusion as general deploys, with stated WTP and fixed gas reservation.
+
```

Comment:
From a technical perspective, I think it's best if we keep `transfers` and `wasm_deploys` separate. It's part of an external interface, and we should maintain the separation there (as is desired in https://casperlabs.atlassian.net/browse/NDRS-735).